File: sentimental-credit/credit.py
import sys
import re


def main():
    card = card_input()
    check_number_digit(card)
    check_luhn(card)


def card_input():
    while True:
        try:
            card = input("Number: ")
            if len(card) > 0 and int(card):
                break
        except ValueError:
            print("INVALID")
    return card


def check_number_digit(card):
    if len(card) not in (13, 15, 16):
        print("INVALID")
        exit()
    else:
        return card


def check_luhn(card):
    digits = [int(x) for x in card]

    # convert digits to select the second-to-last easily
    # second-to-last become second-from-begin
    digits = digits[::-1]

    even = digits[1::2]

    odd = digits[0::2]

    sum_even = 0
    for number in even:
        sum_even += (number * 2 // 10) + (number * 2 % 10)

    sum_odd = 0
    sum_odd += sum(odd)

    total = 0
    total = sum_even + sum_odd

    if total % 10 != 0:
        print("INVALID")
        return

    start_of_card = int(card[:2])

    if start_of_card in (51, 52, 53, 54, 55) and len(card) == 16:
        print("MASTERCARD")
    elif start_of_card in (34, 37) and len(card) == 15:
        print("AMEX")
    elif start_of_card // 10 == 4 and len(card) in (13, 16):
        print("VISA")
    else:
        print("INVALID")


if __name__ == "__main__":
    main()

File: credit/credit.c
#include <cs50.h>
#include <stdio.h>

int main(void)
{
    long number;
    number = get_long("Number: ");
    int from_last = 0;
    int from_second_last = 0;
    int first_two_digit;
    int i = 0;
    while (number > 0)
    {
        if (i % 2 == 0)
        {
            from_last += number % 10;
        }
        else
        {
            int each_one_from_second_last = (number % 10) * 2;
            if (each_one_from_second_last >= 10)
            {
                from_second_last += (each_one_from_second_last % 10) + (each_one_from_second_last / 10);
            }
            else
            {
                from_second_last += each_one_from_second_last;
            }
        }
        if (9 < number && number < 100)
        {
            first_two_digit = number;
        }
        number = number / 10;
        i++;
    }
    int length = i;
    int sum = from_last + from_second_last;
    int last_digit = sum % 10;
    if (last_digit != 0)
    {
        printf("INVALID\n");
    }
    else if (length == 15 && (first_two_digit == 34 || first_two_digit == 37))
    {
        printf("AMEX\n");
    }
    else if (length == 16 && first_two_digit <= 55 && first_two_digit >= 51)
    {
        printf("MASTERCARD\n");
    }
    else if ((length == 16 || length == 13) && (first_two_digit) / 10 == 4)
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }
}